### PR TITLE
SCUMM: Allow replacing Loom's music with audio tracks

### PR DESCRIPTION
**Pull request by eriktorbjorn**
NOTE: This pull request makes changes to the themes. I have not rebuilt the theme ZIP files, nor have I updated the THEMERC files. The themes have changed since I branched, and I want to avoid conflicts.

~~For various reasons, I probably won't have much time to finish this in the coming weeks, so I'm posting this here for feedback, and in case someone wants to take it over.~~

The purpose of this pull request is to allow certain versions of Loom to replace the synthesized music with standalone audio tracks. These tracks should be named the same way as you would normally name them for games that use CD audio. Obviously, we can't distribute such tracks ourselves (unless someone takes it upon themselves to create new ones from the original Swan Lake score, which is surely in the public domain now). But we can write detailed instructions for how to do so, given the appropriate source material. My own notes, which are subject to corrections, so far:

| Track | Usage | FM Towns Track | TurboGrafx-16 Track | Ballet |
|-----|-----|-----|-----|-----|
| 1 | Overture[^1] | 1 or 9 | 14 | Act II, the swans swim on the lake<br>Act II, Odette and the Swans return to the lake |
| 2 | Loom | 2 or 10 | 15 | Act I, Pas de trois, Intrada |
| 3 | Clerics | 3 or 11 | 16 | Act I, Pas de trois, Andante sostenuto |
| 4 | Blacksmiths | 4 or 12 | 17 | Act I, Pas de trois, Moderato |
| 5 | Shepherds / Dragon caves[^2] | 5 or 13 | 18 | Act I, Pas d'action, Dance of the Drunken Courtier Wolfgang |
| 6 | Dragon (abduction)[^2] | 5 or 13 | 18 | Act I, Pas d'action, Dance of the Drunken Courtier Wolfgang |
| 7 | Glassmakers | 6 or 14 | 19 | Act II, Dance of the Swans |
| 8 | Weavers | 7 or 15 | 20 | Act IV, Dance of the Cygnets |
| 9 | Chaos[^3] | 8 or 16 | 21 | Act IV, Finale |
| 10 | Opening the sky | 1 or 9 | 14 | Near the end of the Overture |
| 11 | Chaos Appears / Unmaking the Loom | 8 or 16 | 21 | Near the end of Chaos |

When extracting from the ballet recording, it's probably a good idea to keep the lead-in short ~~but leave a few seconds of silence at the end for looping tracks~~. The music never loops. In fact, some tracks are unlikely to ever play all the way to the end.

[^1]: The Overture appears at least twice in the ballet. Use the one you like best.
[^2]: The FM Towns and TurboGrafx-16 tracks (and, at least on my CD, the ballet) contains both the Shepherds and the Dragon abduction, so you have to cut them apart.
[^3]: Chaos uses part of the Act IV finale. It's unclear how much. The TurboGrafx-16 and FM Towns audio tracks go on for much longer than the Roland MT-32 track. They start at the same point in the music and go on for 4.5 minutes - all the way to the end of the ballet. Use your own judgment. It's unlikely that the game will play the track all the way to the end anyway.

What remains to be done? ~~Well, I've only played the game to Crystalgard (the Glassmakers) myself. That said...~~

- ~~The Overture doesn't work. It depends on a music timer, and that's currently not updated. (The CD timer normally used for audio tracks runs much too quickly, so I'm not using that here.) The text "OVERTURE" fades in when the timer reaches 4, fades out when it reaches 198, and the rest of the intro starts when it reaches 204. Note that the intro then continues with the Overture music. The intro music doesn't start until the Overture music ends. This is good, because it means we don't have to match the length of the original audio track.~~
- ~~The patch currently only supports the EGA and Mac versions of Loom. I have little doubt that at least the Amiga version could be included, but I don't have that one myself.~~
- ~~There are a couple of musical cues that aren't handled, e.g. when the sky is opened. Identifying them in the ballet may or may not be possible. I haven't bothered trying.~~
- ~~I may tweak the timer a bit so that the Overture matches the MT-32 version better. Even though I like how the Overture is timed to the Ozawa version right now. SCUMM ticks probably aren't accurate enough to match the original timer. Maybe the timing of the Overture should be configurable to accommodate for different recordings?~~
- ~~The demos use different IDs for the music. There is code to handle this, but unfortunately there doesn't seem to be any good way to distinguish the demo from the full game at the moment. This is the subject of a different pull request, that I still need help with: https://github.com/scummvm/scummvm/pull/3684~~
- ~~The Overture Timing setting is shown both for the EGA and VGA versions. There doesn't seem to be any good way to distinguish them at this point. Perhaps checking of the file 000.LFL exists in the game path? That _should_ mean it's the VGA version.~~ If you have an old game entry, the Overture Timing setting may still show up for the VGA version. I don't think that's worth fixing, though.

And of course, this also opens the door to using the movie soundtrack for Indiana Jones and the Last Crusade, but I'm not going down that road...

Addendum: It was pointed out to me that the Loom music is probably based on the 1979 Seiji Ozawa recording with the Boston Symphony Orchestra. ~~I do not own this myself, but based on the album I found on Spotify,~~ these are the necessary tracks and - where applicable - approximate times from that recording:

| Game Track | Name | Disc | Track | Time | Title |
|-----|-----|-----|-----|-----|-----|
| 1 | Overture[^1] | 1<br>2 | 17<br>4 | whole track | Act II: No. 10 Scène (Moderato)<br>Act II: No. 14 Scène (Moderato) |
| 2 | Loom | 1 | 5 | whole track | Act I: No. 4a Pas de trois: Intrada (Allegro) |
| 3 | Clerics | 1 | 6 | whole track | Act I: No. 4b Pas de trois: Andante sostenuto |
| 4 | Blacksmiths | 1 | 8 | whole track | Act I: 4d Pas de trois: Moderato |
| 5 | Shepherds / Dragon caves | 1 | 14 | 00:00-2:25 | Act I: No. 6 Pas d'action und No. 7 Sujet |
| 6 | Dragon (abduction) | 1 | 14 | 2:25-2:50 | Act I: No. 6 Pas d'action und No. 7 Sujet |
| 7 | Glassmakers | 1 | 23 | whole track | Act II: No. 13d Danse des petits cygnes (Allegro moderato) |
| 8 | Weavers | 2 | 20 | whole track | Act IV: No. 27 Danses des petits cygnes (Moderato) |
| 9 | Chaos | 2 | 22 | 1:01-2:50[^3] | Act IV: No. 29 Scène finale (Andante) |
| 10 | Opening the sky | 1<br>2 | 17<br>4 | 1:48-1:58<br>1:50-2:01 | Act II: No. 10 Scène (Moderato)<br>Act II: No. 14 Scène (Moderato) |
| 11 | Chaos Appears / Unmaking the Loom | 2 | 22 | 2:42-2:50 | Act IV: No. 29 Scène finale (Andante) |

Not every recording has these as separate tracks, though. E.g. on my On my Naxos CD (see link in comment below) all of the "Pas de trois" tracks are merged into a single track. Even with the Ozawa recording, som assembly is required. Track 10 and 11 are the ones that need the most attention, but all of the tracks are optional. ScummVM will play the original music if a standalone track is missing.

The corresponding list for my Naxos CDs:

| Game Track | Name | Disc | Track | Time | Title |
|-----|-----|-----|-----|-----|-----|
| 1 | Overture[^1] | 1<br>2 | 11<br>1 | whole track | Act II: Scène<br>Act II: Scène |
| 2 | Loom | 1 | 5 | 0:00-2:05 | Act I: Pas de trois |
| 3 | Clerics | 1 | 5 | 2:05-4:42 | Act I: Pas de trois |
| 4 | Blacksmiths | 1 | 5 | 5:57-6:59 | Act I: Pas de trois |
| 5 | Shepherds / Dragon caves | 1 | 7 | 00:00-1:52 | Act I: Pas d'action |
| 6 | Dragon (abduction) | 1 | 7 | 1:52-2:20 | Act I: Pas d'action |
| 7 | Glassmakers | 1 | 14 | 5:24-6:45 | Act II: Dance of the Swans |
| 8 | Weavers | 2 | 14 | whole track | Act IV: Dance of the Cygnets |
| 9 | Chaos | 2 | 16 | 1:18-3:04[^3] | Act IV: Finale |
| 10 | Opening the sky | 1<br>2 | 11<br>1 | 2:01-2:09<br>2:02-2:09 | Act II: Scène<br>Act II: Scène |
| 11 | Chaos Appears / Unmaking the Loom | 2 | 16 | 2:56-3:04 | Act IV: Finale |

### CODE REVIEW

**Comment by eriktorbjorn** [comment]
For reference, here's the part of script-44 that appears to handle the Ouverture:

```
[0039] (08) if (VAR_SOUNDCARD != 4) {
[0040] (2A)   startScript(39,[25],F);
[0046] (18) } else {
[0049] (33)   SetPalColor(0,15)
[004F] (80)   breakHere();
[0050] (07)   setState(898,1);
[0054] (2A)   startScript(39,[24],F);
[005A] (58)   beginOverride();
[005C] (18)   goto 00C2;
[005F] (80)   breakHere();
[0060] (38)   unless (VAR_MUSIC_TIMER >= 4) goto 005F;
[0067] (33)   SetPalColor(1,15)
[006D] (80)   breakHere();
[006E] (33)   SetPalColor(9,15)
[0074] (80)   breakHere();
[0075] (33)   SetPalColor(11,15)
[007B] (80)   breakHere();
[007C] (33)   SetPalColor(15,15)
[0082] (80)   breakHere();
[0083] (38)   unless (VAR_MUSIC_TIMER >= 198) goto 0082;
[008A] (33)   SetPalColor(11,15)
[0090] (80)   breakHere();
[0091] (33)   SetPalColor(9,15)
[0097] (80)   breakHere();
[0098] (33)   SetPalColor(1,15)
[009E] (80)   breakHere();
[009F] (33)   SetPalColor(0,15)
[00A5] (80)   breakHere();
[00A6] (07)   setState(898,0);
[00AA] (80)   breakHere();
[00AB] (33)   SetPalColor(15,15)
[00B1] (80)   breakHere();
[00B2] (38)   unless (VAR_MUSIC_TIMER >= 204) goto 00B1;
[00B9] (72)   loadRoom(2);
[00BB] (32)   setCameraAt(160);
[00BE] (2E)   delay(60);
[00C2] (**) }
```

**Comment by AndywinXp** [comment]
This is awesome, I personally can't help much given that my knowledge of Loom is very limited, but I certainly hope this will be finished sooner or later since it's a very good idea IMHO.

**Comment by eriktorbjorn** [comment]
The Overture track is almost exactly 2.5 minutes long. In that time, the music timer advances to 279, probably close to becoming 280. (Both the FM Towns track and the ballet recording I'm using are about 2:40, but that's close enough to make no difference, in my opinion.)

Given this, we should be able to do a reasonable approximation from how much of the track has been played. But I don't think the CD audio player provides this information at the time of writing, so something will probably have to be done there. At least for the emulated case, where it doesn't actually access a real CD.

The CD player works with "frames" (there are 75 frames per second), so if there's a way to get the current frame that would probably be sufficient. It's just an approximation anyway. The caller can calculate the rest, given the start frame (0 in our case).

**Comment by eriktorbjorn** [comment]
With the latest change, the Loom Overture seems to work fine. I still have to play through the whole game to see that everything's ok. (First impressions are quite good, though.)

**Comment by eriktorbjorn** [comment]
As I understand it, ScummVM recognizes the following versions of Loom:

| Game Version | Affected by this pull request? |
|-----|-----|
| DOS EGA | Yes |
| Macintosh | Yes |
| Amiga | No, but it should be possible |
| Atari ST | No, but it should be possible |
| DOS VGA | No |
| Windows/Mac Steam | No |
| TurboGrafx-16 | No |
| FM Towns | No |

**Comment by bluegr** [comment]
Great work! One question, why are the rest of the versions not affected by this? Do they work in a different manner?

**Comment by eriktorbjorn** [comment]
> Great work! One question, why are the rest of the versions not affected by this? Do they work in a different manner?

The FM Towns and TurboGrafx-16 versions already use audio tracks for the music (the FM Towns version apparently has an alternative track for each song?), so you could theoretically replace them with an orchestral recording already. Though it may not be as easy as I first imagined, because I think some of the tracks on the CD have more than one piece of music (the Shepherds/Dragon track mentioned in the list above), so the scripts probably ask to play the track from a specific timestamp. If so, if your track doesn't match that, it's going to sound bad.

For the Amiga and Atari ST versions, I don't own copies of them so I can't verify if the tracks have the same sound IDs there are in the DOS and Mac versions. I was hoping someone would be able to provide me with that information before declaring this ready to merge, but so far there have been no takers.

**Comment by AndywinXp** [comment]
The best and ideal way to do this would be to have a marker embedded in the audio track (or in a text file with a "map" of the audio track) which once reached triggers a callback which makes the SCUMM script advance. Of course this is easier said than done since it would require a big amount of patching... It's still an idea though.

**Comment by eriktorbjorn** [comment]
> Doing this removes any need for having to wait a specific amount of ticks for each custom version of the track.

I've still only played up to meeting the glassmakers with this patch, but it may be that only the Overture actually needs the music timer. I'm not sure if it waits for any specific part of the music, though. The track I used from my Naxos CD - https://www.naxos.com/catalogue/item.asp?item_code=8.550246-47 - worked well enough. Though I guess I have to try and think of a way to get it to work if the track is too short.

Enhancing the music for the versions that already use CD audio tracks, while it should be possible it's outside the scope of this pull request, and not something I plan on looking at myself.

**Comment by eriktorbjorn** [comment]
There's one more bit of music timer stuff in script 44 that I missed until now:

```
[025F] (48) if (VAR_SOUNDCARD == 4) {
[0266] (28)   if (!Local[9]) {
[026B] (80)     breakHere();
[026C] (38)     unless (VAR_MUSIC_TIMER >= 278) goto 026B;
[0273] (**)   }
[0273] (**) }
[0273] (72) loadRoom(2);
```

So it's important that the music timer reaches 278. I'm in the process to add some code to make this work reasonably even if the track ends prematurely.

**Comment by eriktorbjorn** [comment]
There are at least three more tracks that may be interesting:

| AdLib | Roland | Description |
|-----|-----|-----|
| 19 | 51 | Opening the sky. This can be extracted from late in the Overture |
| 20 | 52 | I don't know what this is used for, but it can be extracted at the end of Shepherds / Dragon cave |
| 21 | 53 | Unmaking the loom. This can be extracted from the end of Chaos |

I'll probably end up adding support for these, though I'd like to see what (if anything) the middle track is used for. If I do so I'll number them 10-12 (even though they have lower sound IDs than the others) to make it clear that they're less important than the others. And I'll update the tables in the description.

Other than that, I think the remaining music are mostly various harp figures. I don't think they're worth bothering with, if they can even be found on their own in the ballet.

**Comment by eriktorbjorn** [comment]
I've played through the EGA version. Hopefully I'll be able to test the Mac version as well, but I don't foresee any major problems. (Though I have a feeling the Mac beeps and boops will be horribly loud compared to the music. Oh well.) So unless anyone has any objections to the implementation, this is pretty much done now.

I'm still hoping someone will be able to tell me if the Amiga and Atari ST versions can be covered by the patch as well.

**Comment by eriktorbjorn** [comment]
The Loom demos will play the wrong music, but that's only because ScummVM doesn't set the GF_DEMO flag for them. I just submitted a separate pull request for that issue: https://github.com/scummvm/scummvm/pull/3684

**Comment by einstein95** [comment]
Looking at the Atari ST version, here's the tracks used:

| Game track | Name |
| ---------- | ---- |
| 1 | Loom |
| 2 | Clerics |
| 3 | Blacksmiths |
| 4 | Dragon (abduction) |
| 5 | Glassmakers |
| 6 | Weavers |
| 7 | Chaos |

**Comment by eriktorbjorn** [comment]
> Looking at the Atari ST version, here's the tracks used:

Thanks! Looks like it's the same as the EGA and Mac versions. I'm not sure, but I think 22 and 23 are used for things like using drafts and learning new notes. There are some harp things like that that I assumed wouldn't be possible to extract cleanly from a ballet recording.

At least not until we find a symphony orchestra filled with fanatic Loom fans who offers to record everything for us for free. ;-)

**Comment by eriktorbjorn** [comment]
I'll enable the replacement tracks for the ST versions later. Once I've heard back about the Amiga version, assuming it too uses the same numbers from the tracks, this bit:

```
	if (_game.id == GID_LOOM) {
		useReplacementAudioTracks = (_game.platform == Common::kPlatformDOS && _game.version == 3) || _game.platform == Common::kPlatformMacintosh;
	}
```

could probably be boiled down to

```
	useReplacementAudioTracks = (_game.id == GID_LOOM && !(_game.features & GF_AUDIOTRACKS));
```

or something like that. That should exclude the TurboGrafx-16 (a.k.a PC Engine), FM Towns and VGA versions, all in one swell foop.

**Comment by eriktorbjorn** [comment]
The only remaining thing I know about now is that the Overture doesn't sync to the intro quite the way the MT-32 version does. I might tweak that a bit, even though I rather like the way the Ozawa version reaches a crescendo as the LucasFilm logo appears now.

File: engines/scumm/sound.cpp [hunk 1]
```diff
@@ -108,6 +111,67 @@ Sound::~Sound() {
...
+	return
+		(_vm->_game.id == GID_LOOM) &&
+		(_vm->_game.version == 3) &&
+		(_vm->_game.platform == Common::kPlatformDOS) &&
+		(_vm->VAR(_vm->VAR_SOUNDCARD) == 4);
+}
+
+// When timing the MT-32 version, it took on average 149.6 seconds for the
+// timer to reach 278. At 60 SCUMM ticks per second, this would be 8976 ticks
+// extept... apparently you can't really use SCUMM ticks to this level of
```
Comment (einstein95) [inline engines/scumm/sound.cpp:124]: minor typo, should be `except`

Comment (eriktorbjorn) [reply engines/scumm/sound.cpp]: > minor typo, should be `except`

Oops. Thanks!

**Comment by eriktorbjorn** [comment]
Here is what a Loom Overture tempo setting might look like: [loom-overture-tempo-setting.txt](https://github.com/scummvm/scummvm/files/7979256/loom-overture-tempo-setting.txt)

But the more I look at it, the less I like it:

- It's tricky to get the setting to only appear on the Engine tab for the appropriate version of the game.
- It's probably hard to sufficiently explain to the user what the setting does without getting overly technical.
- It's probably tricky to get the interval for the setting just right for it to be useful.
- I just can't imagine a lot of users spending time and effort adjusting the timing for an Overture they'll probably skip most of the time.

Also, despite my best efforts, the settings dialog is pretty ugly. I can't even figure out how to get it to use the available space for the slider widget:

scummvm-00003

**Comment by eriktorbjorn** [comment]
After some head-scratching, I've come up with this instead. This is probably good enough to commit, though since it required changes to the theme files I don't want to commit those ZIP files yet. The tempo is now specified relative to the default tempo. Each tick on the slider equals ten SCUMM ticks, and the range is currently set to -200 to 200. Using this, I was able to set the transition in the Overture to happen between approximately 1:30 and 2:20, which should be enough for most sensible recordings of it.

And it should be simple enough to drag the slider back to 0 if you're running ScummVM at a sensible resolution.

scummvm-00004

**Comment by AndywinXp** [comment]
I personally like this solution, good job! One very very minor suggestion: is there a way for the slider to be more intuitive about what it's doing? Like some indication that moving the slider to the right delays the ending point and moving it to the left rushes it. (Or I might have got them backwards, I don't know :P)

**Comment by eriktorbjorn** [comment]
I don't know. I was toying with the idea of adding some text, but I couldn't get it to position right. I may have to describe it in the tooltip, though if that grows too long it looks weird in low resolutions.

**Comment by eriktorbjorn** [comment]
After the latest commit, the tooltip is: "Adjusts the timing of the EGA Loom Overture when using enhanced music. The lower the value, the earlier the Lucasfilm logo appears."

Is that clear enough, or does it need further clarification?

**Comment by AndywinXp** [comment]
Definitely works for me :D thanks

**Comment by Ohrwerk** [comment]
Thank you for this wonderful option, but it seems that not all orchestras are conducted at the same speed, which also provides variety.
540 ticks (2:50 min) for the Overture seems a bit too low to me.
Even on Youtube you can only find overtures with a playing time from 2:53 min upwards.
The overture I used even has a length of not quite 3:03 min , so in my case I would need at least 581 ticks. I have already removed the silence from the beginning and end, I could maybe get another 1.5 sec out of it with a fade at the end, but that's not enough either.

Is it possible to increase the ticks, or is the 540  a magic limit that cannot be exceeded?

**Comment by antoniou79** [comment]
> Thank you for this wonderful option, but it seems that not all orchestras are conducted at the same speed, which also provides variety. 540 ticks (2:50 min) for the Overture seems a bit too low to me. Even on Youtube you can only find overtures with a playing time from 2:53 min upwards. The overture I used even has a length of not quite 3:03 min , so in my case I would need at least 581 ticks. I have already removed the silence from the beginning and end, I could maybe get another 1.5 sec out of it with a fade at the end, but that's not enough either.
> 
> Is it possible to increase the ticks, or is the 540 a magic limit that cannot be exceeded?

The slider setting should not correspond to the full length of the overture track, but to the timestamp _within_ that track when it reaches a sort of crescendo, which is when the first screen of the game proper (the night sky over the sea) appears on-screen, similarily to when playing the original MT-32 track.

**Comment by eriktorbjorn** [comment]
What antoniou79 said. I've uploaded a brief clip of the transition here: https://www.youtube.com/watch?v=kqIMGlm-PZE

But if necessary it is of course easy to extend the range of the slider. See engines/scumm/dialogs.cpp, LoomEgaGameOptionsWidget():

```
	_overtureTicksSlider = new GUI::SliderWidget(widgetsBoss(), "LoomEgaGameOptionsDialog.OvertureTicks", _("When using replacement music, this adjusts the time when the Overture changes to the scene with the Lucasfilm and Loom logotypes."), kOvertureTicksChanged);

	// In the Ozawa recording, the transition happens at about 1:56. At is
	// turns out, this is a fairly fast version of the tune. After checking
	// a number of different recordings, I've settled on an interval of
	// 1:40 - 2:50. This is larger than I had hoped, but I guess it's
	// really necessary.
	//
	// Hopefully that still means you can set the slider back to its default
	// value at most reasonable screen resolutions.

	_overtureTicksSlider->setMinValue(-160);
	_overtureTicksSlider->setMaxValue(540);
```

The values are expressed in 1/10th of seconds away from the Ozawa default of 1:56 (defined in `DEFAULT_LOOM_OVERTURE_TRANSITION`), i.e. the min value is 1:56 - 16 seconds, and the max value is 1:56 + 54 seconds.

So if you let us know where that transition happens in your recording, we could round up to the nearest 10 seconds or so.

**Comment by Ohrwerk** [comment]
Thank you very much for the help.

I calculated wrong with the ticks anyway, I started from 0-540 and not from -160 to 540.
If I play the overture as in the original during the night starry sky, there is no need to "screw" the overture slider.
I just have the problem that Track2 (Loom Song) does not play when the Overture ends and I thought I could fix this problem by "turning" the knob.
Whether Track2 is an Mp3 or the MT32 is used doesn't matter. 
I have now looked at it again, in the original (MT32) after the end of the overture the loom song is played without problems.
Only if I press ESC during the MP3 overture, the game jumps to the Night Starry Sky with the Lucasfilm Games logo and the Loom Song (MT32 or Track2.mp3) is played.

**Comment by eriktorbjorn** [comment]
> I just have the problem that Track2 (Loom Song) does not play when the Overture ends and I thought I could fix this problem by "turning" the knob.
> Whether Track2 is an Mp3 or the MT32 is used doesn't matter.

I can't reproduce that in the current development version. I'll have to try it in the stable release, but I don't have the time right now.

**Comment by eriktorbjorn** [comment]
I've tried letting the Overture play out in its entirety (using a FLAC file as the audio track) now both in 2.7.0 and 2.7.1, and in both the next piece of music started playing as expected.

I'm not saying there isn't a problem, I just don't know where to start looking for it.

**Comment by Ohrwerk** [comment]
Wait, it's probably a misunderstanding on my part. Sorry
More details will follow when I have tested everything.

**Pull request merged**